TPP.Persistence.MongoDB/Repos/PollRepo.cs

```diff
@@ -0,0 +1,132 @@
+        }
+
+        public PollRepo(IMongoDatabase database, IClock clock)
+        {
+            database.CreateCollectionIfNotExists(CollectionName).Wait();
+            Collection = database.GetCollection<Poll>(CollectionName);
+            _clock = clock;
+            InitIndexes();
+        }
+
+        private void InitIndexes()
+        {
+            Collection.Indexes.CreateMany(new[]
+            {
+                new CreateIndexModel<Poll>(Builders<Poll>.IndexKeys.Ascending(u => u.PollCode)),
+                new CreateIndexModel<Poll>(Builders<Poll>.IndexKeys.Ascending(u => u.CreatedAt)),
+            });
+        }
+
+        public async Task<Poll> CreatePoll(
+            string pollCode, string pollTitle, bool multiChoice, bool allowChangeVote,
+            IImmutableList<string> pollOptions)
+        {
+            List<PollOption> pollOptionsObjects = pollOptions
+                .Select((option, index) => new PollOption(
+                    id: index + 1,
+                    option: option,
+                    voterIds: new List<string>()
+                )).ToList();
+
+            Poll poll = new(
+                pollCode: pollCode,
+                pollTitle: pollTitle,
+                voters: new List<string>(),
+                pollOptions: pollOptionsObjects,
+                _clock.GetCurrentInstant(),
+                multiChoice: multiChoice,
+                alive: true,
+                allowChangeVote: allowChangeVote
+            );
+
+            await Collection.InsertOneAsync(poll);
+
+            return poll;
+        }
+
+        public async Task<VoteFailure?> Vote(string pollCode, string userId, IImmutableList<int> options)
+        {
+            Poll? poll = await FindPoll(pollCode);
+            if (poll == null)
+                return new VoteFailure.PollNotFound(pollCode);
+            if (!poll.Alive)
+                return new VoteFailure.PollNotAlive();
+            if (!poll.AllowChangeVote && poll.Voters.Contains(userId))
+                return new VoteFailure.AlreadyVoted();
+            if (options.Count == 0)
+                return new VoteFailure.CannotVoteForNone();
+            if (options.Count > 1 && !poll.MultiChoice)
+                return new VoteFailure.NotMultipleChoice();
+
+            await Collection.UpdateOneAsync(
+                p => p.PollCode == pollCode,
+                Builders<Poll>.Update.AddToSet(p => p.Voters, userId));
+
+            ImmutableList<int> invalidOptions = options.Except(poll.PollOptions.Select(p => p.Id)).ToImmutableList();
+            if (invalidOptions.Any())
+                return new VoteFailure.InvalidOptions(invalidOptions);
+
+            // remove any existing votes in case the user has voted before and is changing their vote right now.
+            // no typed support for the $[] (for each in array) operator yet, see https://jira.mongodb.org/browse/CSHARP-2232
+            foreach (int i in Range(0, poll.PollOptions.Count))
```

Comment:
should only be done if AllowChangeVote is true, otherwise is a waste of time

Reply:
I'd rather keep it for robustness. This code is by no means latency critical